org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/EclipseMavenLauncher.java

```diff
@@ -0,0 +1,118 @@
+/*******************************************************************************
+ * Copyright (c) Christoph Läubrich
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *      Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.m2e.internal.launch;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import org.osgi.service.component.annotations.Component;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.debug.core.DebugException;
+import org.eclipse.debug.core.DebugPlugin;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.debug.core.ILaunchConfigurationType;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.debug.core.ILaunchManager;
+import org.eclipse.debug.core.ILaunchesListener2;
+import org.eclipse.debug.core.model.IProcess;
+import org.eclipse.debug.ui.DebugUITools;
+import org.eclipse.debug.ui.IDebugUIConstants;
+
+import org.eclipse.m2e.actions.MavenLaunchConstants;
+import org.eclipse.m2e.core.internal.launch.IMavenLauncher;
+
+
+/**
+ * EclipseMavenLauncher
+ */
+@Component(service = IMavenLauncher.class)
+public class EclipseMavenLauncher implements IMavenLauncher {
+
+  public CompletableFuture<?> runMaven(File basedir, String goals, Properties mavenProperties) throws CoreException {
+    ILaunchManager launchManager = DebugPlugin.getDefault().getLaunchManager();
+    ILaunchConfigurationType launchConfigurationType = launchManager
+        .getLaunchConfigurationType(MavenLaunchConstants.LAUNCH_CONFIGURATION_TYPE_ID);
+    ILaunchConfigurationWorkingCopy workingCopy = launchConfigurationType.newInstance(null,
+        launchManager.generateLaunchConfigurationName("mavenrun-" + UUID.randomUUID()));
+    workingCopy.setAttribute(MavenLaunchConstants.ATTR_POM_DIR, basedir.getAbsolutePath());
+    workingCopy.setAttribute(MavenLaunchConstants.ATTR_GOALS, goals);
+    workingCopy.setAttribute(MavenLaunchConstants.ATTR_SAVE_BEFORE_LAUNCH, false);
+    List<String> properties = new ArrayList<>();
+    for(String propertyKey : mavenProperties.stringPropertyNames()) {
+      String propertyValue = mavenProperties.getProperty(propertyKey);
+      if(propertyKey != null && propertyKey.trim().length() > 0) {
+        String prop = propertyKey.trim() + "=" + (propertyValue == null ? "" : propertyValue);
```

Comment:
```suggestion
        String prop = propertyKey.trim() + (propertyValue == null ? "" : "=" + propertyValue);
```
Besides that I wonder if there is a realistic chance that the `propertyKey` is null?

Furthermore what do you think about using a `Map<String, String>` instead of `Properties`?

Reply:
`Properties` are actually `Map<String, String>` after generic erasure :-)
As maven itself uses `Properties` (and they are also named like that) it seems a good choice here even though its not that much relevant.